python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -713,6 +733,15 @@ def _submit_runs_and_update_backfill_in_chunks(
                 submitted_partitions
             )
         )
+        if code_unreachable_error_raised:
+            # Code server became unavailable mid-backfill. Rewind the cursor back to the cursor
+            # from the previous iteration, to allow next iteration to reevaluate the same
+            # events.
+            backfill_data_with_submitted_runs = (
+                backfill_data_with_submitted_runs.replace_latest_storage_id(
+                    previous_asset_backfill_data.latest_storage_id
+                )
+            )
```

Comment:
an alternative would be to, rather than updating the storage Id and then potentially un-updating it, waiting until here to update it in the 'expected' case where there is no error (and not doing that if there was an error). That won't work if we use latest_storage_id before this point though and expect it to be the new value

Reply:
This is a good callout.

We're currently doing a weird thing where we call `execute_asset_backfill_iteration_inner`, which returns the expected resultant asset backfill data (included requested partitions and latest storage id) after all runs are submitted. This function is used for testing to assert that a backfill iteration will result in X partitions being requested.

Then because we might have mid-iteration backfill cancellations and code location errors like this happening during run submission, we un-update the requested partitions and the cursor when unexpected things happen.

I think we should refactor `execute_asset_backfill_iteration_inner` to stop doing these un-updates. Instead we could handle observed changes (updated materialized/failed partitions) separately from expected state updates (partitions to request/next cursor), which I can do if I get the chance.

In the meantime, because there's already existing logic to do "un-updates", I'd prefer to keep the cursor "un-update" logic with the existing logic.